Move at::chunk into the graph fuser

zou3519 opened the pull request:
... to avoid slow at::chunk (it is slow due to tensor initialization). Picking up from #10026

This is done through the following:

1) Absorb starting chunks into FusionGroup as a part of the graph fuser
pass.
2) When compiling a kernel, emit a `std::vector<ConcatDesc>` that describes if an input (of the original graph) will be chunked.
3) When launching a kernel, `use std::vector<ConcatDesc>` to chunk an
input tensor on the CPU. This chunk directly takes in an at::Tensor and creates
four TensorInfo structs in-place in the argument list, bypassing the creation of intermediate Tensors.

## Test Plan

- Expect test and correctness test to see if a single chunk is fused
  by the graph fuser
- Correctness test for a variety of chunks (dimension = beginning,
  middle, end) and tensors (contiguous, non-contiguous, edge case
  (splitSize = 1) for both CPU/CUDA
- Expect test for multiple chunks fused into the same kernel and
  correctness test.

cc @zdevito @apaszke 

## Perf
LSTM forward pass, 1 layer, 512 hidden size and input size, 100 seq length, requires_grad=False on all inputs and weights.

After changes:
```
thnn    cudnn   jit
8.8468  6.5797  9.3470
```

Before changes:
```
thnn    cudnn   jit
9.9221  6.6539  11.2550
```

zou3519 committed:
Move at::chunk into the graph fuser

This is done through the following:

1) Absorb starting chunks into FusionGroup as a part of the graph fuser
pass.
2) When compiling a kernel, move chunks out of the FusionGroup and use
the resulting graph. Emit a std::vector<MaybeChunkDesc> that describes
if an input (of the original graph) will be chunked.
3) When launching a kernel, use std::vector<MaybeChunkDesc> to chunk an
input tensor on the CPU. This chunk takes in an at::Tensor and outputs
four TensorInfo structs, bypassing intermediate Tensors.
4) The resulting TensorInfo structs are sent into the compiled kernel.

Test Plan

- Expect test and correctness test to see if a single chunk is fused
  by the graph fuser
- Correctness test for a variety of chunks (dimension = beginning,
  middle, end) and tensors (contiguous, non-contiguous, edge case
  (splitSize = 1) for both CPU/CUDA
- Expect test for multiple chunks fused into the same kernel and
  correctness test.

Absorb starting at::chunk into FusionGroups

If all outputs to at::chunk are inputs to a FusionGroup and "chunks",
"dim" are both constants, then the at::chunk is moved into the beginning
of the FusionGroup.

Teach fusion compiler about at::chunk inside a FusionGroup

When compiling, the compiler emits an extra std::vector<ConcatDesc> that
says which inputs are chunked into how many pieces. The compiler scans
inputs and produces a list of "flat inputs".

When launching, the compiler scans the inputs and the chunk_desc to see
which inputs are chunked. It uses this information to prepare a list of
flat inputs to send to the compiled kernel.

Update expect files

Fix nit

Windows fix

Address most comments, still working on the rest

Use prim::FusedChunk for chunks inside FusionGroup.

Addressed comments:

- add assert
- separate PartitionDesc chunk / cat ctors so the logic is clearer

zdevito commented on torch/csrc/jit/fusion_compiler.h:
```diff
@@ -60,6 +60,15 @@ struct AnnotatedGraph {
   std::vector<TensorDesc> output_desc;
 };
 
+struct ChunkDesc {
```
Now that I think about this more, I believe ChunkDesc and ConcatDesc are actually exactly the same thing. Each describes how to split up a bigger tensor into smaller components that become inputs into the kernel. The only difference is that for chunks it describes an input, and for cat, an output. However from the kernels perspective they are all just pointers to memory anyway. It should cleaner to actually have the code reflect this similarity, and because they are the same concept, they should also be able to share a lot of code. For instance I was suggesting adding a precomputed TensorDesc to Chunk -- well concat already has that and the handling for it.

zou3519 replied:
Good point! I'll look into using ConcatDesc and sharing code between chunk/cat

zdevito commented on torch/csrc/jit/fusion_compiler.cpp:
```diff
@@ -624,27 +641,179 @@ void compressContiguous(
+  return it - start;
+}
+
+// If there are fused Chunk nodes in the graph, returns a new Graph replacing
```
If you look at how this is handled for concat, it takes a more lightweight approach. By leaving the Concat operators in and just skipping them in code generation we can still generate the correct kernel, and make sure the input list has the 'flattened' inputs added, without having to copy/do surgery on the fusion graph. It is possible that something similar might work here (see later comment about how chunk and concat are basically the same exact operator).

zdevito commented on torch/csrc/jit/fusion_compiler.cpp:
```diff
@@ -728,6 +909,13 @@ struct CUDAFusionFunction : public CompiledFusionFunction {
     TORCH_CUDA_CHECK(cudaGetDeviceProperties(&prop, agraph.device));
     checkCUDAVersion(prop);
 
+    // Use a graph without the fused chunk nodes to emit a kernel.
```
It doesn't seem right for implementation details like having fused chunks to leak into the CUDA/CPU backend code and get repeated twice. This will be hard to maintain if every new feature added to the fuser had to do something similar.

zdevito commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -339,7 +339,11 @@ struct GraphFuser {
           JIT_ASSERT(input->node()->kind() == prim::Constant);
           Node * in_const = subgraph.createClone(input->node(), [](Value*) -> Value* { throw std::runtime_error("unexpected input"); });
-          subgraph.prependNode(in_const);
+          if (insert_after == nullptr) {
```
what caused this to need to change?

zdevito commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -349,17 +353,24 @@ struct GraphFuser {
-      group->removeInput(p);
-      subgraph.inputs()[p]->replaceAllUsesWith(in_graph->output());
-      subgraph.eraseInput(p);
+    for (size_t i = 0; i < n->outputs().size(); ++i) {
```
why did this change? multiple outputs can happen due to chunk operators?

zou3519 replied:
Yes, this changed because chunk operators have multiple outputs

zdevito commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -436,6 +448,70 @@ struct GraphFuser {
+      return false;
+    }
+    // Does the chunk have constant chunks/dim?
+    if (!isConstant(chunk->namedInput(attr::chunks)) ||
```
`chunk->is_constant(attr::chunks)`

zdevito reviewed (commented):
The fusion pass changes look good. The fusion compiler stuff looks pretty good too but I have some suggestions to simplify it and to avoid allocations along the fast path that I think we should do. Otherwise it would make it difficult to add more functionality to the fuser later.

apaszke commented on torch/csrc/jit/fusion_compiler.cpp:
```diff
@@ -624,41 +668,94 @@ void compressContiguous(
+// XXX: this code assumes that inputs are 32-bit addressable
+static inline uint32_t computeNumel(at::ArrayRef<int64_t> sizes) {
+  uint32_t result = 1;
+  if (sizes.size() == 0) {
```
Isn't that the scalar case? Shouldn't it return `1` as numel?

apaszke commented on torch/csrc/jit/fusion_compiler.cpp:
```diff
@@ -624,41 +668,94 @@ void compressContiguous(
+  at::IntList map_size;
+  uint32_t numel;
+  std::vector<int64_t> keep_alive_size;
+  if (chunk_desc[0].nSubtensors == 1) {
```
Didn't you assert this never happens above?

zou3519 commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -610,6 +682,10 @@ struct GraphFuser {
+    // Fuse starting chunks into the group.
+    for (auto it = block->nodes().rbegin(); it != block->nodes().rend();) {
+      it = scanNodeForChunks(*it);
+    }
```
Yes I set an invariant that fused chunks can only take in inputs to fusion groups.

zou3519 commented on torch/csrc/jit/fusion_compiler.cpp:
```diff
@@ -445,11 +458,39 @@ std::pair<std::vector<ConcatDesc>, bool> emitCompilationUnit(
+    for(auto p : subgraph.inputs()) {
+      if (auto chunk = maybeIsChunk(p)) {
+        int64_t dim = toIValue(chunk->namedInput(attr::dim))->toInt();
+        int64_t chunks = toIValue(chunk->namedInput(attr::chunks))->toInt();
```
I haven't thought much about it but I will think about it

zou3519 replied:
Oh I see, `prim::FusedConcat` has `dim` as an attribute on the node. Yes, that does look nicer, I'll give it a shot

zou3519 commented:
This should be ready for another review now, despite the failing (unrelated) tests

apaszke commented on test/expect/TestJit.test_fusion_distribute.expect:
```diff
@@ -1,20 +1,16 @@
-  %6 : Float(4!, 2), %7 : Float(4!, 2) = aten::chunk(%1, %3, %2)
-  %8 : Float(4, 2) = prim::FusionGroup_0[device=0](%4, %6, %5, %7)
-  return (%8);
+  %2 : Float(4, 2) = prim::FusionGroup_0[device=0](%0, %1)
```
This test is for distributing chunk over pointwise ops we could fuse. Is that transformation still a thing? We should be able to fuse all chunks anyway, right? Did you remove it?

zou3519 replied:
We still distribute the chunks before we fuse any chunks in. Fusing the chunks in first prevents possible fusion opportunities (because the invariant for prim::FusedChunk is that only inputs to the FusionGroup can be inputs to prim::FusedChunk).

This expect files shows that the chunk was distributed over the fusible pointwise ops, then everything was fused into one fusion group

apaszke commented on torch/csrc/jit/fusion_compiler.cpp:
```diff
@@ -418,7 +418,18 @@ std::string encodeRHS(Node * n) {
 
-std::pair<std::vector<ConcatDesc>, bool> emitCompilationUnit(
+static at::optional<Node&> usedInFusedChunk(Value * input) {
+  // If input is an input to prim::FusedChunk, it will only have one use.
```
Assert plz

apaszke commented on torch/csrc/jit/fusion_compiler.h:
```diff
@@ -60,24 +60,37 @@ struct AnnotatedGraph {
-      // so dim - 1 is no longer contiguous
-      cont[dim - 1] = false;
+    if (is_chunk) {
+      subtensorDesc.reset(new TensorDesc(desc.scalar_type, desc.contiguity));
```
Are you sure that we don't need to modify the contiguity similarly to what we do below? Chunking along any dimension other than the first one will change a contiguous input to non-contiguous chunks.

```python
In [2]: import torch

In [3]: x = torch.randn(10, 10)

In [4]: x.chunk(10, 0)[0].is_contiguous()
Out[4]: True

In [5]: x.chunk(10, 1)[0].is_contiguous()
Out[5]: False
```

zou3519 replied:
Yes, `desc.contiguity` should already have the correct contiguity:
https://github.com/pytorch/pytorch/blob/60b58f153f0325eeb585d030da9af61f124b4aa7/torch/csrc/jit/fusion_compiler.cpp#L470-L476 
That part of the code constructs `desc` with the correct sizes / strides of each of the output chunks, and then the TensorDesc constructor computes the correct contiguity. `desc` is then passed into the PartitionDesc ctor. 

This behavior is different from what happens in a `cat` operation: for `aten::cat` PartitionDesc computes the correct contiguity, but for `aten::chunk` PartitionDesc is passed the correct contiguity. I'll add comments, a separate constructor, or move some code around as appropriate to make this distinction clearer.

apaszke reviewed (approved):
LGTM, but I'm not sure if you're handling contiguity information correctly in `PartitionDesc`

zou3519 committed:
Fix bug in prim::FusedChunk

If one has a graph like the following:
```
y1, y2 = chunk(x)
z1, z2 = chunk(x)
fusiongroup(y1, y2, z1, z2)
```
Only one chunk should become a prim::FusedChunk inside the fusion group
because there is an invariant that prim::FusedChunk inside the fusion
group may not have the same input. This is because of how the fusion
compiler replaces the input to be chunked into its chunked tensors.
